google-cloud-pubsub/lib/google/cloud/pubsub/flow_controller.rb

```diff
@@ -0,0 +1,159 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+require "google/cloud/pubsub/errors"
+require "concurrent/atomics"
+
+module Google
+  module Cloud
+    module PubSub
+      ##
+      # @private
+      #
+      # Used to control the flow of messages passing through it.
+      #
+      class FlowController
+        attr_reader :message_limit
+        attr_reader :byte_limit
+        attr_reader :limit_exceeded_behavior
+        ##
+        # @private Implementation accessors
+        attr_reader :outstanding_messages, :outstanding_bytes, :awaiting_message_acquires, :awaiting_bytes_acquires
+
+        def initialize message_limit: 1000, byte_limit: 10_000_000, limit_exceeded_behavior: :ignore
+          # init MonitorMixin
+          # super()
+          unless [:ignore, :error, :block].include? limit_exceeded_behavior
+            raise ArgumentError, "limit_exceeded_behavior must be one of :ignore, :error, :block"
+          end
+          @mutex = Mutex.new
+          @message_limit = message_limit
+          @byte_limit = byte_limit
+          @limit_exceeded_behavior = limit_exceeded_behavior
+          @outstanding_messages = 0
+          @outstanding_bytes = 0
+
+          @awaiting_message_acquires = []
+          @awaiting_bytes_acquires = []
+        end
+
+        def acquire message_size
+          return if limit_exceeded_behavior == :ignore
+          @mutex.lock
+          if limit_exceeded_behavior == :error && @outstanding_messages + 1 > message_limit
+            raise FlowControlLimitError, "Flow control message limit (#{message_limit}) would be exceeded"
+          end
+          if limit_exceeded_behavior == :error && @outstanding_bytes + message_size > byte_limit
+            raise FlowControlLimitError,
+                  "Flow control byte limit (#{byte_limit}) would be exceeded, message_size: #{message_size}"
+          end
+          if limit_exceeded_behavior == :block && message_limit < 1
+            raise FlowControlLimitError,
+                  "Flow control message limit (#{message_limit}) exceeded by a single message, would block forever"
+          end
+          if limit_exceeded_behavior == :block && message_size > byte_limit
+            raise FlowControlLimitError,
+                  "Flow control byte limit (#{byte_limit}) exceeded by a single message, would block forever"
+          end
+
+          acquire_message
+          acquire_bytes message_size
+        ensure
+          @mutex.unlock if @mutex.locked?
```

Comment:
This should be `@mutex.owned?` rather than `@mutex.locked?`. The latter still returns true if the mutex is locked by a different thread; the former returns true only if it is locked by the current thread.

Reply:
Thank you for catching this. Fixed!